accounts/store.go

```diff
@@ -0,0 +1,311 @@
+	"github.com/lightningnetwork/lnd/lnwire"
+	"go.etcd.io/bbolt"
+)
+
+const (
+	// DBFilename is the filename within the data directory which contains
+	// the macaroon stores.
+	DBFilename = "accounts.db"
+
+	// dbFilePermission is the default permission the account database file
+	// is created with.
+	dbFilePermission = 0600
+
+	// dbPathPermission is the default permission the account database
+	// directory is created with (if it does not exist).
+	dbPathPermission = 0700
+
+	// DefaultAccountDBTimeout is the default maximum time we wait for the
+	// account bbolt database to be opened. If the database is already
+	// opened by another process, the unique lock cannot be obtained. With
+	// the timeout we error out after the given time instead of just
+	// blocking for forever.
+	DefaultAccountDBTimeout = 5 * time.Second
+)
+
+var (
+	// accountBucketName is the name of the bucket where all accounting
+	// based balances are stored.
+	accountBucketName = []byte("accounts")
+
+	// lastAddIndexKey is the name of the key under which we store the last
+	// known invoice add index.
+	lastAddIndexKey = []byte("last-add-index")
+
+	// lastSettleIndexKey is the name of the key under which we store the
+	// last known invoice settle index.
+	lastSettleIndexKey = []byte("last-settle-index")
+
+	// byteOrder is the binary byte order we use to encode integers.
+	byteOrder = binary.BigEndian
+)
+
+// BoltStore wraps the bolt DB that stores all accounts and their balances.
+type BoltStore struct {
+	db *bbolt.DB
+}
+
+// NewBoltStore creates a BoltStore instance and the corresponding bucket in the
+// bolt DB if it does not exist yet.
+func NewBoltStore(dir, fileName string) (*BoltStore, error) {
+	// Ensure that the path to the directory exists.
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err := os.MkdirAll(dir, dbPathPermission); err != nil {
+			return nil, err
+		}
+	}
+
+	// Open the database that we'll use to store the primary macaroon key,
+	// and all generated macaroons+caveats.
+	path := filepath.Join(dir, fileName)
+	db, err := bbolt.Open(path, dbFilePermission, &bbolt.Options{
+		Timeout: DefaultAccountDBTimeout,
+	})
+	if err == bbolt.ErrTimeout {
+		return nil, fmt.Errorf("error while trying to open %s: timed "+
+			"out after %v when trying to obtain exclusive lock",
+			path, DefaultAccountDBTimeout)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// If the store's bucket doesn't exist, create it.
+	err = db.Update(func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(accountBucketName)
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Return the DB wrapped in a BoltStore object.
+	return &BoltStore{db: db}, nil
+}
+
+// Close closes the underlying bolt DB.
+func (s *BoltStore) Close() error {
+	return s.db.Close()
+}
+
+// NewAccount creates a new OffChainBalanceAccount with the given balance and a
+// randomly chosen ID.
+func (s *BoltStore) NewAccount(balance lnwire.MilliSatoshi,
```

Comment:
Are we assuming the account manager is always holding a mutex here? Or that a random ID collision will be caught when we got to update the account itself?